### PR TITLE
ENH: EADtype._find_compatible_dtype

### LINKED ISSUE #24246: EA interface: add maybe_promote?
I'm preparing a PR that addresses the short-comings of `core.dtypes.cast.maybe_promote` uncovered in #23982, and there's currently only a branch in there along the following lines:
```
    def maybe_promote(dtype, fill_value)
    [...]
    elif is_extension_array_dtype(dtype) and isna(fill_value):
        fill_value = dtype.na_value
    [...]
    return dtype, fill_value
```

This does not describe any promotion rules, and (like the rest of `maybe_promote`) is very inconsistent about the returned missing value (see #23982).

My suggestion would be to to something like the following:

### LINKED ISSUE #32346: BUG: assignment with enlargement gives object dtype with ExtensionArrays
From https://github.com/pandas-dev/pandas/issues/32271, when setting with enlargment, the dtype gets converted into object dtype (while for normal integer or boolean dtype this is not the case):

```
In [10]: s = pd.Series([1, 2, 3], dtype="Int64")  

In [11]: s[3] = 4 

In [12]: s   
Out[12]: 
0    1
1    2
2    3
3    4
dtype: object

In [13]: s = pd.Series([1, 2, 3], dtype="int64") 

In [14]: s[3] = 4   

In [15]: s  
Out[15]: 
0    1
1    2
2    3
3    4
dtype: int64
```

It also happens with eg DecimalArray from our tests, so suppose it is a general issue with ExtensionArrays.

### LINKED ISSUE #45349: ENH: ExtensionDtpye._dtype_with_na/_maybe_promote
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd

ii = pd.interval_range(1, 10)
ser = pd.Series(ii).copy()
mask = np.zeros(ser.shape, dtype=bool)
mask[1] = True

ii.insert(1, np.nan)  # <- IntervalDtype[float64]
ii.putmask(mask, np.nan)  # <- IntervalDtype[float64]

ser.where(mask, np.nan)  # <- ValueError
ser[1] = np.nan  <- coerces to object
```

### Issue Description

ATM dtypes.cast.ensure_dtype_can_hold_na is incorrect for EA dtypes that cannot hold NA values (the IntervalDtype[int] example is the only one that comes to mind).

We have special-casing for IntervalDtype in Index._find_common_type_compat which is why the Index methods above behave well.  Short-term, we can move that special-casing into ensure_dtype_can_hold_na.  Longer-term, we need a way for EADtype subclasses to specify this behavior.

This could be seen as a special case of xref #24246

cc @jorisvandenbossche @TomAugspurger 

### Expected Behavior

a

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### LINKED ISSUE #52235: BUG: Adding a pyarrow scalar value with the same type to a series series changes the dtype into object
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
s = pd.Series(datetime.now(), dtype='date32[pyarrow]')

s[1] = pa.scalar(datetime.now(), type='date32')

s
Out[60]: 
0    2023-03-27
1    2023-03-27
dtype: object
```

```
s = pd.Series([1,2,3], dtype='int32[pyarrow]')

s[3] = pa.scalar(4, type='int32')

s
Out[72]: 
0    1
1    2
2    3
3    4
dtype: object
```
```


### Issue Description

Adding new pyarrow scalar value of the same type to a series changes its dtype into object. This make it impossible(?) to add to a date32 series without changing its dtype.

### Expected Behavior

The series dtype should be kept.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 5c155883fdc5059ee7e21b20604a021d3aa92b01
python           : 3.11.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19043
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_Indonesia.1252

pandas           : 2.1.0.dev0+284.g5c155883fd
numpy            : 1.24.2
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.6.3
pip              : 23.0.1
Cython           : 0.29.32
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.2
html5lib         : 1.1
pymysql          : None
psycopg2         : 2.9.5
jinja2           : 3.1.2
IPython          : 8.6.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : 1.3.7
brotli           : 1.0.9
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : 2.8.4
odfpy            : None
openpyxl         : 3.1.0
pandas_gbq       : None
pyarrow          : 11.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : 1.4.45
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
zstandard        : None
tzdata           : 2022.6
qtpy             : 2.3.0
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #24246
- [x] closes #45349
- [x] closes #32346
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Tests copied from #52833, so this may close #52235.

Needs a lot of cleanup, in particular docs.  The ArrowDtype._find_compatible_dtype implementation is quite kludgy, would appreciate help from someone more knowledgeable.

I think `_find_compatible_dtype` is the "correct" long-term non-kludge solution for setitem-with-expansion.

### CODE REVIEW

File: pandas/core/arrays/arrow/dtype.py [hunk 1]
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
         array_class = self.construct_array_type()
         arr = array.cast(self.pyarrow_dtype, safe=True)
         return array_class(arr)
+
+    def _maybe_promote(self, item: Any) -> tuple[DtypeObj, Any]:
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/dtype.py:330]: Shouldn't this be implemented this on `ExtensionDtype` too?

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/dtype.py:330]: it is https://github.com/pandas-dev/pandas/pull/53106/files#diff-2e345cb935fc3bae33ad469d4187e9607f679f19e30afb6d764f7c7c4f32639dR394

Comment (mroeschke) [reply pandas/core/arrays/arrow/dtype.py:330]: Ah gotcha. I didn't notice that. Would be good for the base function to have a docstring (as you mentioned in the OP) so it's clear what this method is supposed to do

File: pandas/core/arrays/arrow/dtype.py [hunk 2]
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
...
+        if isinstance(item, pa.Scalar):
+            if not item.is_valid:
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/dtype.py:332]: Shouldn't NA always be able to be inserted into ArrowExtensionArray?

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/dtype.py:332]: im not sure.  pyarrow nulls are typed, so we could plausibly want to disallow e.g. `<pyarrow.TimestampScalar: None>` in a pyarrow integer dtype

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -2091,26 +2086,14 @@ def _setitem_with_indexer_missing(self, indexer, value):
...
-                if not is_object_dtype(self.obj.dtype):
-                    # Every NA value is suitable for object, no conversion needed
-                    value = na_value_for_dtype(self.obj.dtype, compat=False)
-
-                new_dtype = maybe_promote(self.obj.dtype, value)[0]
-
-            elif isna(value):
-                new_dtype = None
+            new_dtype = None
+            if is_list_like(value):
```
Comment (mroeschke) [inline pandas/core/indexing.py:2090]: Note for ArrowDtype with `pa.list_` type, we would want to treat `value` like a scalar e.g

```
ser = pd.Series([[1, 1]], dtype=pd.ArrowDtype(pa.list_(pa.int64())))
ser[0] = [1, 2]
```

File: pandas/core/arrays/arrow/dtype.py [hunk 4]
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
...
+                if item.type == self.pyarrow_dtype:
+                    return self, item.as_py()
+                if item.type.to_pandas_dtype() == np.int64 and self.kind == "i":
+                    # FIXME: kludge
+                    return self, item.as_py()
+
+            item = item.as_py()
+
+        elif item is None or item is libmissing.NA:
+            # TODO: np.nan? use is_valid_na_for_dtype
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/dtype.py:343]: Since pyarrow supports nan vs NA, possibly we want to allow nan if `pa.types.is_floating(self.pyarrow_dtype)`

File: pandas/core/arrays/arrow/dtype.py [hunk 5]
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
...
         arr = array.cast(self.pyarrow_dtype, safe=True)
         return array_class(arr)
+
+    def _maybe_promote(self, item: Any) -> tuple[DtypeObj, Any]:
+        if isinstance(item, pa.Scalar):
+            if not item.is_valid:
+                # TODO: ask joris for help making these checks more robust
+                if item.type == self.pyarrow_dtype:
+                    return self, item.as_py()
+                if item.type.to_pandas_dtype() == np.int64 and self.kind == "i":
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/dtype.py:336]: Why is this needed specifically?

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/dtype.py:336]: This was just to get the tests copied from #52833 passing.

File: pandas/core/arrays/arrow/dtype.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/arrow/dtype.py:343]: what to do here depends on making a decision about when/how to distinguish between np.nan and pd.NA (which i hope to finally nail down at the sprint).  doing this The Right Way would involve something like implementing EA._is_valid_na_for_dtype

File: pandas/core/dtypes/dtypes.py [hunk 6]
```diff
@@ -2281,3 +2302,29 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
         array_class = self.construct_array_type()
         arr = array.cast(self.pyarrow_dtype, safe=True)
         return array_class(arr)
+
+    def _maybe_promote(self, item: Any) -> tuple[DtypeObj, Any]:
+        if isinstance(item, pa.Scalar):
+            if not item.is_valid:
+                # TODO: ask joris for help making these checks more robust
```
Comment (jbrockmendel) [inline pandas/core/dtypes/dtypes.py:2309]: @jorisvandenbossche any thoughts here?  (not time-sensitive)

**Comment by jbrockmendel** [comment]
Updated with some added tests, changed EADtype._maybe_promote->_find_compatible_dtype.

There are some TODOs mostly for pyarrow and related to nan-vs-NA, but I don't think any of them are blockers.

**Comment by jbrockmendel** [comment]
_find_compatible_dtype can be used to implement a default implementation of _is_valid_scalar_for, which can be used to implement _from_scalars (both discussed on this week's dev call).

File: pandas/core/dtypes/dtypes.py [hunk 7]
```diff
@@ -2326,3 +2347,29 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
         array_class = self.construct_array_type()
         arr = array.cast(self.pyarrow_dtype, safe=True)
         return array_class(arr)
+
+    def _find_compatible_dtype(self, item: Any) -> tuple[DtypeObj, Any]:
+        if isinstance(item, pa.Scalar):
```
Comment (phofl) [inline pandas/core/dtypes/dtypes.py:2352]: what happens if item is null? The pyarrow null

File: pandas/tests/extension/test_arrow.py [hunk 8]
```diff
@@ -2754,6 +2754,29 @@ def test_describe_timedelta_data(pa_type):
     tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.parametrize(
+    "value, target_value, dtype",
+    [
+        (pa.scalar(4, type="int32"), 4, "int32[pyarrow]"),
+        (pa.scalar(4, type="int64"), 4, "int32[pyarrow]"),
+        # (pa.scalar(4.5, type="float64"), 4, "int32[pyarrow]"),
```
Comment (phofl) [inline pandas/tests/extension/test_arrow.py:2762]: What happens here?

Also what happens with a int64 scalar and int32 dtype?

File: pandas/core/dtypes/dtypes.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:2352]: IIUC pyarrow nulls are now typed.  id prefer to be strict about making these match, but dont care _that_ much.  am hoping @jorisvandenbossche will weigh in

File: pandas/tests/extension/test_arrow.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/tests/extension/test_arrow.py:2762]: id want to follow the same logic we do for numpy dtypes, but was punting here in expectation of doing it in a follow-up (likely involving joris expressing an opinion)
